api/envoy/config/cluster/v3/cluster.proto

```diff
@@ -812,6 +821,14 @@ message Cluster {
   // of 0 would indicate that none of the timeout was used or that the timeout was infinite. A value
   // of 100 would indicate that the request took the entirety of the timeout given to it.
   bool track_timeout_budgets = 47;
+
+  // Optional customization and configuration of upstream type.
+  // By default, Envoy will create an upstream of type envoy.filters.upstreams.http.default
+  // The default upstream will create a envoy.filters.upstreams.http.tcp upstream
+  // if the request is a CONNECT request and connect termination is configured.
+  // Otherwise it will be envoy.filters.upstreams.http.http, with the protocol
+  // based on `http2_protocol_options`.
+  UpstreamType upstream_type = 48;
```

Comment:
In terms of field name, I'm torn between being more specific here versus keeping this generic because I think ultimately we would want to use that for tcp_proxy also? Should the docs be more clear about which filters this applies to currently? This is what I was talking about re: cluster.proto being a mess. There are a huge amount of concepts convolved into one place and it's very confusing trying to understand what fields are used by what. I'm not saying we can fix this without a major version bump / major deprecation spree, but I think we probably need a lot more docs for this type of thing.

Reply:
I think we can reuse the proto for tcp proxy, but we'd have to use the different type if that's allowable.  Basically for tcp proxy, it'd have to be a envoy.filters.upstreams.tcp since the interfaces are all different.
I'll comment that for today, this only works for http upstreams but is designed to be extended to custom tcp upstreams as well.  SG?

Reply:
Right I agree that is how it would have to work, we just need to make sure it's super clearly documented, both here and probably also in the filter docs.